core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +98,164 @@ where
-    pub fn merkle_root(self) -> H256 {
+    pub fn merkle_root(&self) -> H256 {
         if self.hashes.is_empty() {
             let depth = tree_depth_by_size(self.binary_tree_size);
-            self.hasher.empty_subtree_hash(depth)
+            if self.head_index == 0 {
+                self.hasher.empty_subtree_hash(depth)
+            } else {
+                self.left_cache[depth]
+            }
         } else {
-            self.compute_merkle_root_and_path(0, None)
+            self.compute_merkle_root_and_path(0, None, None)
         }
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        let (mut left_path, mut right_path) = (vec![], vec![]);
+        let root_hash =
+            self.compute_merkle_root_and_path(index, Some((&mut left_path, &mut right_path)), None);
+        (root_hash, right_path)
+    }
+
+    /// Returns the root hash and the Merkle proofs for an interval of leafs.
+    /// The interval is [0, `length`), where `0` is the leftmost uncached leaf.
+    pub fn merkle_root_and_paths_for_interval(
+        &self,
+        length: usize,
+    ) -> (H256, Vec<H256>, Vec<H256>) {
+        let (mut left_path, mut right_path) = (vec![], vec![]);
+        let root_hash = self.compute_merkle_root_and_path(
+            length - 1,
+            Some((&mut left_path, &mut right_path)),
+            None,
+        );
+        (root_hash, left_path, right_path)
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.hashes.push_back(leaf_hash);
+        if self.head_index + self.hashes.len() > self.binary_tree_size {
+            self.binary_tree_size *= 2;
+        }
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn cache(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        self.compute_merkle_root_and_path(count - 1, None, Some(&mut new_cache));
+        self.hashes.drain(0..count);
+        self.head_index += count;
+        self.left_cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
```

Comment:
This code looks quite confusing and difficult to maintain.

- I'd suggest to push the elements from the cached path to `hashes` as the first element once you've reached an appropriate level. In this way, you eliminate quite a few corner cases; e.g., `hashes` will always start with an even index.
- I may misunderstand what `cache` is, but isn't it essentially a Merkle path for the first retained tree element? (Technically, it can be a subset of this path to the left of the retained element.) If so, why do you need a separate argument / computation for this path, rather than getting it as a left path for the first retained element?